barretenberg/cpp/src/barretenberg/boomerang_value_detection/graph.cpp

```diff
@@ -0,0 +1,672 @@
+    for (const auto& variable_index : ultra_circuit_constructor.real_variable_index) {
+        variables_gate_counts[variable_index] = 0;
+        variables_degree[variable_index] = 0;
+        variable_adjacency_lists[variable_index] = {};
+    }
+
+    std::map<FF, uint32_t> constant_variable_indices = ultra_circuit_constructor.constant_variable_indices;
+    const auto& arithmetic_block = ultra_circuit_constructor.blocks.arithmetic;
+    auto arithmetic_gate_numbers = arithmetic_block.size();
+    bool arithmetic_gate_exists = arithmetic_gate_numbers > 0;
+    if (arithmetic_gate_exists) {
+        for (size_t i = 0; i < arithmetic_gate_numbers; i++) {
+            auto gate_variables = get_arithmetic_gate_connected_component(ultra_circuit_constructor, i);
+            connect_all_variables_in_vector(ultra_circuit_constructor, gate_variables, false);
+        }
+    }
+    const auto& elliptic_block = ultra_circuit_constructor.blocks.elliptic;
+    auto elliptic_gate_numbers = elliptic_block.size();
+    bool elliptic_gates_exist = elliptic_gate_numbers > 0;
+    if (elliptic_gates_exist) {
+        for (size_t i = 0; i < elliptic_gate_numbers; i++) {
+            std::vector<uint32_t> gate_variables = get_elliptic_gate_connected_component(ultra_circuit_constructor, i);
+            connect_all_variables_in_vector(ultra_circuit_constructor, gate_variables, false);
+        }
+    }
+    const auto& range_block = ultra_circuit_constructor.blocks.delta_range;
+    auto range_gates = range_block.size();
+    bool range_gates_exists = range_gates > 0;
+    if (range_gates_exists) {
+        std::vector<uint32_t> sorted_variables;
+        for (size_t i = 0; i < range_gates; i++) {
+            auto current_gate = get_sort_constraint_connected_component(ultra_circuit_constructor, i);
+            if (!current_gate.empty()) {
+                sorted_variables.insert(sorted_variables.end(), current_gate.begin(), current_gate.end());
+            } else {
+                connect_all_variables_in_vector(ultra_circuit_constructor, sorted_variables, true);
+                sorted_variables.clear();
+            }
+        }
+    }
+
+    const auto& lookup_block = ultra_circuit_constructor.blocks.lookup;
+    auto lookup_gates = lookup_block.size();
+    bool lookup_gates_exists = lookup_gates > 0;
+    if (lookup_gates_exists) {
+        for (size_t i = 0; i < lookup_gates; i++) {
+            std::vector<uint32_t> variable_indices = get_plookup_gate_connected_component(ultra_circuit_constructor, i);
+            connect_all_variables_in_vector(ultra_circuit_constructor, variable_indices, false);
+        }
+    }
+}
+
+template <typename FF>
+bool Graph_<FF>::check_is_not_constant_variable(bb::UltraCircuitBuilder& ultra_circuit_builder,
+                                                const uint32_t& variable_index)
+{
+    // this method checks that variable is not constant, so it's not in constant_variable_indices from ultra circuit
```

Comment:
Grammar: so -> i.e.

Also these comments are better to be above the function. Something like /** */

Reply:
It's done